build system: add list of features for documentation and sanity checking

maribu opened the pull request:
### Contribution description

This PR adds a `features.yaml` file as machine readable but easy to edit format that lists, groups, and documents all features ("feature" as in the the build system entity) in RIOT. A small python script generates a Makefile for and markdown file from that.

The Makefile is used to check if all features provided and requested are actually valid names or typos.

The Markdown file is used by Doxygen to generation a documentation page containing all features.

### Testing procedure

1. Try requesting or providing a non-existing feature. A build for an app+board combo that contains a feature provided or requested that does not exist should fail.
    - Beware: This requires `make generate-features` to be run after touching `features.yaml`, as the generated Makefile is tracked as source. This avoids running a python script before each and every of the current 141584 build targets (which would result in 4h of additional CPU time per CI build, if each build would add 0.1s of CPU time.)
2. Check the generated documentation.
    - Here the markdown for consumption by Doxygen is generated as part of `make doc`, so this would work right away
3. Check that `make static-tests` catches when the list of features in `makefiles/features_existing.inc.mk` is stale (e.g. you updated `features.yaml` but haven't run `make generate-features` yet).

### Issues/PRs references

- [x] Requires https://github.com/RIOT-OS/riotdocker/pull/247

maribu commented:
This also needs some documentation. The quick guide is:

1. Edit `features.yaml` (in the root of the repo)
2. Run `make generate-features` (in the root of the repo)
3. Run `make doc` (in the root of the repo)

maribu commented on doc/doxygen/src/feature_list.md:
```diff
@@ -0,0 +1,584 @@
+| `esp_spiffs`                      | A Serial Peripheral Interface Flash File System can be used.                  |
+| `esp_wifi`                        | An ESP WiFi radio is present.                                                 |
+| `esp_wifi_ap`                     | ESP WiFi SoftAP support is present.                                           |
+| `esp_wifi_enterprise`             | The ESP WiFi inteface supports WPA2 enterprise mode.                          |
```
```suggestion
| `esp_wifi_enterprise`             | The ESP WiFi interface supports WPA2 enterprise mode.                         |
```

maribu commented on doc/doxygen/src/feature_list.md:
```diff
@@ -0,0 +1,584 @@
+
+## Arduino Form Factor Features
+
+Theses features indicate compatibility with a specific form factor, e.g. Arduino UNO or Adafruit Feather
```
```suggestion
These features indicate compatibility with a specific form factor, e.g. Arduino UNO or Adafruit Feather
```

maribu commented on doc/doxygen/src/feature_list.md:
```diff
@@ -0,0 +1,584 @@
+| `periph_rtc_mem`                  | The RTC peripheral provides storage memory for deep sleep. Unlike `backup_ram`, this memory is not mapped in the address space and requires calls to @ref rtc_mem_read and @ref rtc_mem_write to access. |
+| `puf_sram`                        | The @ref sys_puf_sram module can be used to harvest entropy from uninitialized SRAM on cold boot to seed PRNGs.  |
+
+# Bluetooth Low Engery Features
```
```suggestion
# Bluetooth Low Energy Features
```

maribu commented on features.yaml:
```diff
@@ -0,0 +1,821 @@
+        esp_wifi_ap:
+          ESP WiFi SoftAP support is present.
+        esp_wifi_enterprise:
+          The ESP WiFi inteface supports WPA2 enterprise mode.
```
```suggestion
          The ESP WiFi interface supports WPA2 enterprise mode.
```

maribu commented on features.yaml:
```diff
@@ -0,0 +1,821 @@
+      arduino_spi: Indicates that Arduino SPI bus mappings are provided.
+      arduino_uart: Indicates that Arduino UART device mappings are provided.
+  Arduino Form Factor Features:
+    help: Theses features indicate compatibility with a specific form factor,
```
```suggestion
    help: These features indicate compatibility with a specific form factor,
```

maribu commented on features.yaml:
```diff
@@ -0,0 +1,821 @@
+      The @ref sys_puf_sram module can be used to harvest entropy from
+      uninitialized SRAM on cold boot to seed PRNGs.
+
+Bluetooth Low Engery Features:
```
```suggestion
Bluetooth Low Energy Features:
```

maribu commented on features.yaml:
```diff
@@ -0,0 +1,821 @@
+        open and solder bridges SB4, SB5 and SB6 are closed instead, SPI2 is
+        connected to the STMmod+/Pmod connector. Request this feature to use
+        SPI2 with this board configuration.
+
```
```suggestion
```

maribu reviewed (commented):
CI comments

mguetschow commented:
> This also needs some documentation. The quick guide is:
> 
>     1. Edit `features.yaml` (in the root of the repo)
> 
>     2. Run `make generate-features` (in the root of the repo)
> 
>     3. Run `make doc` (in the root of the repo)

Would it make sense to have `doc` depend on `generate-features` then? And we probably should document that workflow right in `features.yaml` to prevent people changing that without running `make generate-features` afterwards (does YAML even have comments?).

mguetschow reviewed (changes requested):
Great work, that's a big step towards better documentation!

Did you write all the descriptions anew or did you take them from Kconfig files? If the latter is the case, we have to be careful of having several sources of truth in the tree and should converge to one eventually. Probably worth discussing at the VMA next week (and maybe wait with this PR until then?).

Some nit's below.

maribu commented:
> This also needs some documentation.

I added the missing documentation now

maribu commented on makefiles/dependency_resolution.inc.mk:
```diff
@@ -68,6 +68,18 @@ else
                           "$(DEPRECATED_MODULES_USED)" 1>&2)
   endif
 
+  # Detect provided / used / optional features that do not exist
```
Yes, indeed. I added this except for `FEATURES_USED`, as this is in intersection of `FEAUTRES_PROVIDED` and the other. So if it is not in any of these super sets, it won't be in the intersection.

maribu commented:
> Did you write all the descriptions anew or did you take them from Kconfig files?

A bit of both. The KConfig files did not have the grouping and since we disabled the feature check in the CI has started to bit-rot.

> If the latter is the case, we have to be careful of having several sources of truth in the tree and should converge to one eventually. Probably worth discussing at the VMA next week (and maybe wait with this PR until then?).

We did so last VMA: https://forum.riot-os.org/t/virtual-maintainer-assembly-vma-2023-11/4058/6

There was a bit of misunderstanding whether we drop all of KConfig or just the dependency (e.g. features and modules modeling). But in either case, the KConfig features are just a legacy as of now and this tries to salvage what I can and wire this back up into the CI.

maribu commented:
> Would it make sense to have `doc` depend on `generate-features` then?

I've added calling the script to generate the markdown file to `make doc`.

Note: The target `make generate-features` is still there and will no only create the Makefile. The reason I don't like to add this to very build is CI time. I added a test to `static-test` that will check if `features_existing.inc.mk` is still up to date and fail if not. That is not super elegant, but it reduced the CI overhead a lot and should just work.

maribu commented:
May I squash?

MrKevinWeiss reviewed (dismissed):
Do you mind waiting until tomorrow so I can take a quick look?  If it is urgent then you may dismiss.

maribu commented:
> Do you mind waiting until tomorrow so I can take a quick look? If it is urgent then you may dismiss.

No problem. There are still some open questions to be discussed anyway.

MrKevinWeiss reviewed (commented):
Really nice that you are working on this...

Given the constraints (ie. grouping/nesting) probably using just make would not be so nice... maybe one can use wildcards and namespaces etc. but make would not be the best technology for that.

With kconfig, we can, theoretically nest with `menu` but extraction and manipulation of that data would not be as simple as loading a yaml file.

The only thing I would say about the yaml structure would be something like:
```yaml
- title: Architecture Features
  help: These indicate architecture features, such as word size of the CPU,
        supported instruction sets and so on. All architecture features provided
        by a given board will always be used, e.g. an 8-bit CPU cannot just stop
        being an 8-bit CPU on request.
  features:
    arch_8bit:
      CPU has a 8-bits architecture
    arch_16bit:
      CPU has a 16-bits architecture

- title: CPU Features
  help: These features are related to CPU capabilities or just used to
        indicated which CPU family is used.
  groups:
    - title: CPU Capabilities
      help: These correspond to features/capabilities provided by certain CPUs
      features:
        generic_cpu_feature:
          A generic CPU feature that is not specific to any CPU family.
      groups:
        - title: Cortex M Specific Features
          help: These features are only available on (some) ARM Cortex M MCUs
          features:
            cpu_check_address: The @ref cpu_check_address can be used to check if
                              accessing a given address would cause a bus fault.
            cortexm_svc:
              ARM Cortex-M Supervisor Calls are available.
        - title: nRF Capabilities
          help: These features are only available on (some) nordic nRF MCUs
          features:
            vdd_lc_filter_reg1:
              An LC filter for use with the internal DC/DC converter is
              present. If this is the case, the DC/DC converter is used over the
              LDO regulator for improved power efficiency.
            vdd_lc_filter_reg0:
              The MCU supports a high voltage supply via an two-stage regulator
              and the board has an LC filter required for using the two-stage
              DC/DC converter. This enables the two-stage DC/DC converter.
```

Though this does not prevent matching titles (I don't expect that to be an issue though).

I would wait for the VMA and maybe just get a consensus on introducing a new structure but I won't block it if people really want it in.

chrysn reviewed (commented):
As requested in https://github.com/RIOT-OS/RIOT/issues/20393, it would be practical to link from the feature to supported boards. https://github.com/RIOT-OS/RIOT/pull/20395 has the infrastructure for obtaining that information (which is currently also generated in the `generate-features` target). To be useful, such lists should make use of board metadata (so that an output can be "all STM32 boards" where applicable), and possibly be collapsed (like, listing 3 and an HTML summary that tells how many there are and expands to the full litany). Given the interdependencies, I suggest not doing this here, but getting this merged and then building on it.

This looks good to me from a high-level perspective (haven't gone through every detail yet, trusting one of those two who have will given a more well-founded ACK that I could give). @mguetschow, have your concerns been addressed?

mguetschow reviewed (approved):
Since we agreed on going forward with this approach for feature documentation at today's VMA, I give an ACK on the implementation as is. As for the specific format, maybe you want to consider @MrKevinWeiss proposal to have fixed keys (moving the title to the sub-structure)?

I do think that the now duplicated information in the Kconfig should be removed in a timely fashion, if not in this PR, then in a follow-up. In the latter case, let's open an issue for it to keep it on the list (maybe you can directly give some pointers as to where you've extracted the information from).

kaspar030 commented:
> I'd vote for that, too. My main reason would be that it's easier to represent in Rust :). ... laze uses that scheme, too.

I'd even go further:

```
groups:
- title: CPU Features
  help: These features are related to CPU capabilities or just used to
        indicated which CPU family is used.
  groups:
    - title: CPU Capabilities
      help: These correspond to features/capabilities provided by certain CPUs
      features:
        - name: generic_cpu_feature
          help: A generic CPU feature that is not specific to any CPU family.
```

This maps cleanly to sth like:

```Rust
struct FeatureYaml {
  groups: Vec<Group>,
}

struct Group {
  name: String,
  help: String,
  groups: Vec<Group>,
  features: Vec<Feature>,
}

struct Feature {
 name: String,
 help: String,
}
```

While not using the yaml map key as name and yaml map value as text gives up the free uniqueness checks, in practice that's trivial to add. And having explicit fields for e.g., "name" and "help" (instead of directly using key&value) allows adding fields.

maribu commented:
Rebased on top of https://github.com/RIOT-OS/RIOT/pull/20408 to pass the CI, and squashed.

chrysn commented:
There's an implementation detail that could yet be improved but could also turn out to be a very deep time sink hole: Instead of (ab)using make with non-file-name targets to regenerate files, we could embrace make and use it to ensure that whenever the generated features_existing.inc.mk is out of date, it gets rebuilt. The reason I think this may be a huge time sink is that that could interact badly with having the file checked in, and the alternative to not having the file checked in is having a post-checkout step that generates it before forking off into thousands of build tasks, which is yet again complicated. So let's not slow things down here with that.

Given there is an ACK and that everything looks discussed and ready, pushing the button.

maribu commented:
Yeah! Thx :)

benpicco commented on features.yaml:
```diff
@@ -0,0 +1,881 @@
+    - name: periph_flashpage_pagewise
+      help: The Flashpage peripheral supports pagewise writing.
+    - name: periph_flashpage_rwee
+      help: The Flashpage peripheral is of the Read While Write.
```
I still don't know what that means :smile:

chrysn commented:
These docs are not new in this PR -- could you open an issue for clarification and assign recent contributors (and PR reviewers) of that feature?
